Add `onSelectedCellChange` event

phanect opened the pull request:
Closes #3090

This PR adds ~~`onCellSelected`~~ `onSelectedCellChange` event to the `<DataGrid>` component.
You can get column index, row index, row object and column object when the cell is selected by click, double click, right-click, Page{Up|Down|Left|Right} key, tab key, or any other actions.

```html
    <DataGrid
      ...
      onSelectedCellChange={ ({ rowIdx, row, column }) => console.log(rowIdx, row, column) }
    />
```

phanect commented:
I renamed the event name from `onCellSelect` to `onCellSelected` since it sounds more natural English. (Let me know if it is not natural.)

phanect commented:
Rebased.

phanect commented:
Rebased.

phanect commented:
@amanmahajan7 @nstepien Sorry for ping, but is there any problems on this PR? I'm happy to fix them if there are.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,162 @@ describe('Events', () => {
+      col2: 'a1'
+    });
+    expect(column?.key).toBe('col1');
+  });
```
I think we can combine all these tests into one test
```suggestion
    // should call onCellSelected when cell is double clicked
    await userEvent.dblClick(getCellsAtRowIndex(0)[1]);
  });
```

phanect replied:
You are right. I have combined the tests.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,162 @@ describe('Events', () => {
+        }}
+      />
+    );
+    await userEvent.pointer({ target: getCellsAtRowIndex(0)[0], keys: '[MouseLeft]' });
```
```suggestion
    await userEvent.pointer({ target: getCellsAtRowIndex(0)[0], keys: '[MouseRight]' });
```
no?

phanect replied:
Oops... Thanks for fixing.

amanmahajan7 commented on README.md:
```diff
@@ -198,6 +198,17 @@ A number defining the height of summary rows.
 
 ###### `onCellKeyDown?: Maybe<(args: CellKeyDownArgs<R, SR>, event: CellKeyboardEvent) => void>`
 
+###### `onCellSelected?: Maybe<(args: CellSelectArgs<R, SR>) => void>;`
+
+Triggered when a cell is selected by click, double click, right click, Page{Up|Down|Left|Right} key, Tab key, or any other actions.
```
```suggestion
###### `onSelectedCellChange?: Maybe<(args: CellSelectArgs<R, SR>) => void>;`

Triggered when the selected cell is changed.
```

phanect commented:
@amanmahajan7 Thanks for the review! I have fixed the problems.

phanect commented on src/DataGrid.tsx:
```diff
@@ -688,16 +692,28 @@ function DataGrid<R, SR, K extends Key>(
       setSelectedPosition({ ...position, mode: 'SELECT' });
     }
+
+    if (onCellSelected && !samePosition) {
```
`row` is used before this `if` block, so I only moved `column`.

phanect commented on src/DataGrid.tsx:
```diff
@@ -688,16 +692,28 @@ function DataGrid<R, SR, K extends Key>(
 
+    const row = rows[position.rowIdx];
+    const column = columns[position.idx];
+    const samePosition = isSamePosition(selectedPosition, position);
```
`row` is used before this `if` block, so I only moved `column`.

phanect commented on src/DataGrid.tsx:
```diff
@@ -166,6 +167,8 @@ export interface DataGridProps<R, SR = unknown, K extends Key = Key> extends Sha
   onCellContextMenu?: Maybe<(args: CellClickArgs<R, SR>, event: CellMouseEvent) => void>;
   onCellKeyDown?: Maybe<(args: CellKeyDownArgs<R, SR>, event: CellKeyboardEvent) => void>;
+  /** Function called whenever a cell is selected */
+  onCellSelected?: Maybe<(args: CellSelectArgs<R, SR>) => void>;
```
OK, I have renamed!

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,70 @@ describe('Events', () => {
     expect(getCellsAtRowIndex(0)[1]).toHaveAttribute('aria-selected', 'true');
   });
+
+  it('should call onSelectedCellChange when cell is clicked', async () => {
```
```suggestion
  it('should call onSelectedCellChange when cell selection is changed', async () => {
```

phanect replied:
Updated in 6633ef12.

amanmahajan7 commented on src/types.ts:
```diff
@@ -190,6 +190,13 @@ export type CellKeyDownArgs<TRow, TSummaryRow = unknown> =
   | EditCellKeyDownArgs<TRow, TSummaryRow>;
 
+export interface CellSelectArgs<TRow, TSummaryRow = unknown> {
+  idx: number;
+  rowIdx: number;
+  row: TRow;
+  column: CalculatedColumn<TRow, TSummaryRow>;
```
```suggestion
  rowIdx: number;
  row: TRow;
  column: CalculatedColumn<TRow, TSummaryRow>;
```
We can easily get `idx` from `column`

phanect replied:
Removed in b7037d1 and 4612a95.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,70 @@ describe('Events', () => {
+    expectOnCellSelectedFired();
+
+    // Selected by double click
+    await userEvent.dblClick(getCellsAtRowIndex(0)[1]);
```
```suggestion
    await userEvent.dblClick(getCellsAtRowIndex(0)[0]);
```

Let's change the selected cell for each action.

phanect replied:
To combine the test, the selected cell has to be the same through all actions to select (click, double click, right-click, arrow key, and tab key). To select a cell by tab key, the selected cell cannot be (0, 0) because you can only move to right cell by tab key. The column has to be 1.

We can add arguments to `expectOnCellSelectedFired()` to change selected cells by each actions. However, according to some experts, I hear you should make the test code less DRY because making the test code too complex causes bugs in test code. Therefore, we should avoid adding the arguments to `expectOnCellSelectedFired()`, in my opinion.

However, I don't have strong opinion on this, so let me know if you prefer to add arguments to `expectOnCellSelectedFired()` so that we can select different cells by the actions.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,70 @@ describe('Events', () => {
+  });
+
+  it('should call onSelectedCellChange only once when cell is double clicked', async () => {
+    let timesToBeCalled = 0;
```
Can use spy
```suggestion
    const onSelectedCellChange = vi.fn();
```

phanect replied:
Updated in 1dd2cef.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,70 @@ describe('Events', () => {
+    render(<EventTest onSelectedCellChange={() => timesToBeCalled++} />);
+    await userEvent.dblClick(getCellsAtRowIndex(0)[0]);
+
+    expect(timesToBeCalled).toBe(1);
```
```suggestion
    expect(onSelectedCellChangeSpy).toHaveBeenCalledTimes(1);
```

phanect replied:
I updated the test, but I'm not familiar with spying, so can you check if I have implemented as you expected?

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,70 @@ describe('Events', () => {
   });
+
+  it('should call onSelectedCellChange when cell is clicked', async () => {
+    let idx: number;
+    let rowIdx: number;
+    let row: Row;
+    let column: CalculatedColumn<Row> | undefined;
```
Can use the same approach here

```
onSelectedCellChangeSpy.toHaveBeenCalledWith(...);
```

phanect replied:
Fixed in 1dd2cef.

phanect commented:
@amanmahajan7 Thanks, I have fixed the issues.

I rebased the commit history by `main`, but I have never changed the existing commits (before [doc: simplify document](https://github.com/adazzle/react-data-grid/pull/3328/commits/0540bb170732c3818208a346a41cf68bfbc7c956)).
The new commits are [doc: improve English expressions](https://github.com/adazzle/react-data-grid/pull/3328/commits/5d8a4415abf5606aa643dec3e7cf8912ee2a10e3) and newer ones.

phanect commented:
@amanmahajan7 Sorry, I forgot [This comment](https://github.com/adazzle/react-data-grid/pull/3328#discussion_r1321626956).

phanect commented on src/DataGrid.tsx:
```diff
@@ -166,6 +167,8 @@ export interface DataGridProps<R, SR = unknown, K extends Key = Key> extends Sha
   /** Function called whenever a cell is right clicked */
   onCellContextMenu?: Maybe<(args: CellClickArgs<R, SR>, event: CellMouseEvent) => void>;
   onCellKeyDown?: Maybe<(args: CellKeyDownArgs<R, SR>, event: CellKeyboardEvent) => void>;
+  /** Function called whenever a cell is selected */
```
Updated in 6633ef12.

phanect commented:
@amanmahajan7 I have made an update for  [this comment](https://github.com/adazzle/react-data-grid/pull/3328#discussion_r1321626956). Now this PR is ready to be reviewed.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,62 @@ describe('Events', () => {
+          col1: 1,
+          col2: 'a1'
+        },
+        column: expect.objectContaining({
+          idx: 1,
+          key: 'col2'
+        })
+      });
+    };
+
+    // Selected by click
+    await userEvent.click(getCellsAtRowIndex(0)[1]);
+    expectOnCellSelectedFired();
+
+    // Selected by double click
+    await userEvent.dblClick(getCellsAtRowIndex(0)[1]);
+    expectOnCellSelectedFired();
+
+    // Selected by right-click
+    await userEvent.pointer({ target: getCellsAtRowIndex(0)[1], keys: '[MouseRight]' });
+    expectOnCellSelectedFired();
+
+    // Selected by ←↑→↓ keys
+    await userEvent.click(getCellsAtRowIndex(1)[1]);
+    await userEvent.keyboard('{ArrowUp}');
+    expectOnCellSelectedFired();
+
+    // Selected by tab key
+    await userEvent.click(getCellsAtRowIndex(0)[0]);
+    await userEvent.keyboard('{Tab}');
+  });
+
+  it('should call onSelectedCellChange only once when cell is double clicked', async () => {
+    const onSelectedCellChange = vi.fn();
+
+    render(<EventTest onSelectedCellChange={onSelectedCellChange} />);
+    await userEvent.dblClick(getCellsAtRowIndex(0)[0]);
+
+    expect(onSelectedCellChange).toHaveBeenCalledTimes(1);
+  });
```
```suggestion
  it('should call onSelectedCellChange when cell selection is changed', async () => {
    const onSelectedCellChange = vi.fn();

    render(<EventTest onSelectedCellChange={onSelectedCellChange} />);

    expect(onSelectedCellChange).not.toHaveBeenCalled();

    // Selected by click
    await userEvent.click(getCellsAtRowIndex(0)[1]);
    expect(onSelectedCellChange).toHaveBeenCalledWith({
      column: expect.objectContaining(columns[1]),
      row: rows[0],
      rowIdx: 0
    });
    expect(onSelectedCellChange).toHaveBeenCalledTimes(1);

    // Selected by double click
    await userEvent.dblClick(getCellsAtRowIndex(0)[0]);
    expect(onSelectedCellChange).toHaveBeenCalledWith({
      column: expect.objectContaining(columns[0]),
      row: rows[0],
      rowIdx: 0
    });
    expect(onSelectedCellChange).toHaveBeenCalledTimes(2);

    // Selected by right-click
    await userEvent.pointer({ target: getCellsAtRowIndex(1)[0], keys: '[MouseRight]' });
    expect(onSelectedCellChange).toHaveBeenCalledWith({
      column: expect.objectContaining(columns[0]),
      row: rows[1],
      rowIdx: 1
    });
    expect(onSelectedCellChange).toHaveBeenCalledTimes(3);

    // Selected by ←↑→↓ keys
    await userEvent.keyboard('{ArrowUp}');
    expect(onSelectedCellChange).toHaveBeenCalledWith({
      column: expect.objectContaining(columns[0]),
      row: rows[0],
      rowIdx: 0
    });
    expect(onSelectedCellChange).toHaveBeenCalledTimes(4);

    // Selected by tab key
    await userEvent.keyboard('{Tab}');
    expect(onSelectedCellChange).toHaveBeenCalledWith({
      column: expect.objectContaining(columns[1]),
      row: rows[0],
      rowIdx: 0
    });
    expect(onSelectedCellChange).toHaveBeenCalledTimes(5);
  });
});
```

We can simplify it

phanect commented on README.md:
```diff
@@ -198,6 +198,17 @@ A number defining the height of summary rows.
+
+Arguments:
+
+- `args.idx`: `number` - column index
```
Oops... I forgot to remove it. Thanks.

phanect commented:
@amanmahajan7 Thanks, I have committed your suggestions.

amanmahajan7 commented on test/events.test.tsx:
```diff
@@ -118,11 +118,65 @@ describe('Events', () => {
+    });
+    expect(onSelectedCellChange).toHaveBeenCalledTimes(5);
+  });
+});
```
```suggestion
```

phanect replied:
Sorry, I missed this.

amanmahajan7 reviewed (approved):
Thanks for your contribution.
